lib/chef/http/authenticator.rb

```diff
@@ -104,6 +152,70 @@ def load_signing_key(key_file, raw_key = nil)
         raise Chef::Exceptions::InvalidPrivateKey, msg
       end
 
+      # takes no parameters. Checks for the password in the registry and returns it if there, otherwise returns false
+      def self.get_cert_password
+        @win32registry = Chef::Win32::Registry.new
+        path = "HKEY_LOCAL_MACHINE\\Software\\Progress\\Authentication"
+        # does the registry key even exist?
+        present = @win32registry.get_values(path)
+        if present.nil? || present.empty?
+          raise Chef::Exceptions::Win32RegKeyMissing
+        end
+
+        if present.map { |h| h[:name] }[0] == "PfxPass"
+          present.map { |h| h[:data] }[0].to_s
+        end
+      rescue Chef::Exceptions::Win32RegKeyMissing
+        # if we don't have a password, log that and generate one
+        Chef::Log.warn "Authentication Hive and value not present in registry, creating it now"
+        new_path = "HKEY_LOCAL_MACHINE\\Software\\Progress\\Authentication"
+        unless @win32registry.key_exists?(new_path)
+          @win32registry.create_key(new_path, true)
+        end
+        # cspell:disable-next-line
+        some_chars = "~!@#%^&*_-+=`|\\(){}[<]:;'>,.?/0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZ_abcdefghijklmnopqrstuvwxyz".each_char.to_a
+        password = some_chars.sample(1 + rand(some_chars.count)).join[0...14]
+        values = { name: "PfxPass", type: :string, data: password }
+        @win32registry.set_value(new_path, values)
+        password
+      end
+
+      def self.retrieve_certificate_key(client_name)
+        require "openssl" unless defined?(OpenSSL)
+
+        if ChefUtils.windows?
+
+          password = get_cert_password
+
+          unless password
+            return false
+          end
+
+          powershell_code = <<~CODE
+              Try {
+                $my_pwd = ConvertTo-SecureString -String "#{password}" -Force -AsPlainText;
+                $cert = Get-ChildItem -path cert:\\LocalMachine\\My -Recurse | Where-Object { $_.Subject -match "#{client_name}" } -ErrorAction Stop;
+                $tempfile = [System.IO.Path]::GetTempPath() + "export_pfx.pfx";
+                Export-PfxCertificate -Cert $cert -Password $my_pwd -FilePath $tempfile;
+              }
+              Catch {
+                return $false
+              }
+          CODE
+          my_result = powershell_exec!(powershell_code).result
+
+          if my_result != false
+            pkcs = OpenSSL::PKCS12.new(File.binread(my_result["PSPath"].split("::")[1]), password)
+            ::File.delete(my_result["PSPath"].split("::")[1])
+            OpenSSL::PKey::RSA.new(pkcs.key.to_pem)
```

Comment:
you can probably explictly `return` this value and then remove the `else false` conditionals and at the end of the function have a one line statement `false` for the error return case.